### PR TITLE
NIFI-1975 - Processor for parsing evtx files

### PR DESCRIPTION
**Pull request by brosander**

### CODE REVIEW

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+            .name("original")
+            .description("The unmodified input FlowFile will be transferred to this relationship")
+            .build();
+
+    @VisibleForTesting
+    static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILURE, REL_ORIGINAL, REL_BAD_CHUNK)));
+
+    @VisibleForTesting
+    static final PropertyDescriptor GRANULARITY = new PropertyDescriptor.Builder().required(true)
+            .name("Granularity")
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: We are now trying to respect a standard by using the combination of .name() and .displayName() on such objects. Discuss here: https://mail-archives.apache.org/mod_mbox/nifi-dev/201605.mbox/%3C5A6FDF1E-1889-46FE-A3C4-5D2F0A905979@apache.org%3E

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(GRANULARITY));
+
+    private final FileHeaderFactory fileHeaderFactory;
+    private final MalformedChunkHandler malformedChunkHandler;
+    private final RootNodeHandler rootNodeHandler;
+    private final XMLStreamWriterFactory xmlStreamWriterFactory;
+    private final ResultProcessor resultProcessor;
+
+    public ParseEvtx() {
+        this(FileHeader::new, ParseEvtx::handleMalformedChunkException, XmlBxmlNodeVisitor::new,
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: You may want to stick to Java 7 or you will need to have a separate PR if we want to merge it in 0.x branch

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java (hunk 1 above)
Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: Thanks! Reading up on that now.

**Comment by pvillard31** [comment]
I am running out of time right now, but will have a deeper look later and give it a try asap. As a general comment, I'd say, even if I know how boring it can be, that some comments in the code would ease the reading ;) anyway, it looks like a nice feature!

**Comment by brosander** [comment]
@pvillard31 thanks for the review so far!  I'll try and beef up the comments a bit.  I think the Java 8 features make for a much cleaner implementation so I'd like to keep them in for now.

(Would have needed 20+ factory classes [here alone](https://github.com/brosander/nifi/blob/NIFI-1502-rebase-pr/nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/parser/bxml/ValueNode.java#L72))

**Comment by joewitt** [comment]
regarding Java8 angle obviously that just means this can only go in 1.x line and not 0.x.  That is ok though.

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE [hunk 3]
```diff
@@ -0,0 +1,35 @@
+nifi-evtx-nar
+Copyright 2016 The Apache Software Foundation
+
+This product includes the following work from python-evtx (https://github.com/williballenthin/python-evtx):
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE]: lines 4-8 are not needed here.  This is necessary for the NOTICE of the source release which is the top level nifi/NOTICE file but this NOTICE file within the nar is only for the binary distribution of the nar and that does not contain these sources.

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Created by brosander on 5/24/16.
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: Recommend removal of these 'Created by brosander' lines on all things.  By contributing the software becomes of the Apache NiFi community.  For provenance considerations necessary for discussion/etc.. we have git blame/etc.. but for all rights/attribution/etc.. we have the Apache License and NOTICE associated with the project overall.

Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: will do, and will configure ide not to add them in the future

File: NOTICE [hunk 5]
```diff
@@ -7,3 +7,7 @@ The Apache Software Foundation (http://www.apache.org/).
 This product includes the following work from the Apache Hadoop project:
 
 BoundedByteArrayOutputStream.java adapted to SoftLimitBoundedByteArrayOutputStream.java
+
+This product includes the following work from python-evtx (https://github.com/williballenthin/python-evtx):
```
Comment (joewitt) [inline NOTICE]: I would recommend rephrasing to 

This contains includes derived works from the Apache Software License V2 library python-evtx (https://github.com/williballenthin/python-evtx) 
  Copyright 2012, 2013 Willi Ballenthin <william.ballenthin@mandiant.com>
      while at Mandiant <http://www.mandiant.com>
The derived work is adapted from Evtx/Evtx.py, Evtx/BinaryParser.py,Evtx/Nodes.py and can be found in the org.apache.nifi.processors.evtx.parser package.

**Comment by joewitt** [comment]
ok i reviewed the handling of the ported python evtx library.  I provided my suggestions on what I believe would be correct.  Given that it was apache licensed making derived works is fortunately pretty straightforward.  While it did not contain any NOTICE and thus items to carry forward it did contain copyright entries on their source files themselves so the recommendations I've made honor those by giving them an honorable mention in the notice file of our source distribution.

And actually now that I think of it you will want to take that rephrased NOTICE section i mention and also include it in the nifi-assembly/NOTICE file.

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE (hunk 3 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE]: i take that back.  They are needed.  But would rephrase to

This contains includes derived works from the Apache Software License V2 library python-evtx (https://github.com/williballenthin/python-evtx) 
Copyright 2012, 2013 Willi Ballenthin william.ballenthin@mandiant.com
while at Mandiant http://www.mandiant.com
The derived work is adapted from Evtx/Evtx.py, Evtx/BinaryParser.py,Evtx/Nodes.py and can be found in the org.apache.nifi.processors.evtx.parser package.

**Comment by brosander** [comment]
Added the notice to the root and assembly, updated it in the nar

Added some comments to the code

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE [hunk 6]
```diff
@@ -0,0 +1,36 @@
+nifi-evtx-nar
+Copyright 2016 The Apache Software Foundation
+
+This includes derived works from the Apache Software License V2 library python-evtx (https://github.com/williballenthin/python-evtx)
+Copyright 2012, 2013 Willi Ballenthin william.ballenthin@mandiant.com
+while at Mandiant http://www.mandiant.com
+The derived work is adapted from Evtx/Evtx.py, Evtx/BinaryParser.py, Evtx/Nodes.py, Evtx/Views.py and can be found in the org.apache.nifi.processors.evtx.parser package.
+
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE]: Seems comprehensive, thanks! I will defer to @joewitt if this is sufficient

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/pom.xml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            <artifactId>nifi-processor-utils</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-mock</artifactId>
+            <scope>test</scope>
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/pom.xml]: Nit pick if there are other comments; group test dependencies together at the bottom

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+    private final XMLStreamWriterFactory xmlStreamWriterFactory;
+    private final ResultProcessor resultProcessor;
+
+    public ParseEvtx() {
+        this(FileHeader::new, ParseEvtx::handleMalformedChunkException, XmlBxmlNodeVisitor::new,
+                ParseEvtx::createWriter, ParseEvtx::processResult);
+    }
+
+    @VisibleForTesting
+    ParseEvtx(FileHeaderFactory fileHeaderFactory, MalformedChunkHandler malformedChunkHandler, RootNodeHandler rootNodeHandler,
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: I imagine the package-local scope on this and other methods are based on the IDE recommendations? I wonder if we should keep them more public in case sub-classing makes sense?

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+        if (basename.endsWith(EVTX_EXTENSION)) {
+            return basename.substring(0, basename.length() - EVTX_EXTENSION.length());
+        } else {
+            logger.warn("Trying to parse file without .evtx extension {} from flowfile {}", new Object[]{basename, flowFile});
+            return basename;
+        }
+    }
+
+    @VisibleForTesting
+    static void processResult(ProcessSession session, ComponentLog logger, FlowFile updated, Exception exception, String basename, ChunkHeader chunkHeader, Record record) {
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: The parameters would seem to be attached to instances, does this need to be a static method?

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+    @VisibleForTesting
+    static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILURE, REL_ORIGINAL, REL_BAD_CHUNK)));
+
+    @VisibleForTesting
+    static final PropertyDescriptor GRANULARITY = new PropertyDescriptor.Builder().required(true)
+            .name("granularity")
+            .displayName("Granularity")
+            .description("Output flow file for each Record, Chunk, or File encountered in the event log")
+            .allowableValues(RECORD, CHUNK, FILE)
+            .build();
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: With allowable values, usually there is a required(true) and something like defaultValue(RECORD). Otherwise there should be a validator or documentation to say the default value used (if not supplied)

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/pom.xml (hunk 7 above)
Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/pom.xml]: will do

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java (hunk 8 above)
Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: was just pulling out bits of logic to make testing each chunk easier, I can make it more visible if necessary

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java (hunk 9 above)
Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: it doesn't need to be, all of those parameters are initialized in onTrigger so they aren't really scoped to the object, I figured it'd be easier to test this logic as a static method

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java (hunk 10 above)
Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/main/java/org/apache/nifi/processors/evtx/ParseEvtx.java]: there is a requried(true) on the end of the first line there, I'll make that more obvious with formatting, probably default to granularity of chunk

**Comment by brosander** [comment]
@mattyb149 I think I've addressed your suggestions, please let me know if further changes are needed

**Comment by mattyb149** [comment]
The nifi-evtx-nar needs to be added to the top-level POM and the nifi-assembly POM, otherwise it will not be included in the distro.

**Comment by brosander** [comment]
@mattyb149 updated those poms, verified that the nar is in the assembly

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE (hunk 6 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-nar/src/main/resources/META-INF/NOTICE]: i am hardly authoritative but i did review this case and believe it to be correct.  Some would argue it is more than necessary i'm sure but let's err on the side of doing more than we must.

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/test/java/org/apache/nifi/processors/evtx/ParseEvtxTest.java [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+import static org.mockito.Mockito.anyString;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.isA;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ParseEvtxTest {
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/test/java/org/apache/nifi/processors/evtx/ParseEvtxTest.java]: This class tests the individual methods in the ParseEvtx processor, but not the processor lifecycle (like onTrigger). Can you add some more tests that exercise the processor? An example of using the nifi-mock framework can be found in TestEvaluateXPath, it has the TestRunner stuff with flowfiles, relationships, asserts, etc. You will likely want a test file or two to be used as input, although if line endings/whitespace are important in the format you may just need the data directly in the Java code.

File: nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/test/java/org/apache/nifi/processors/evtx/ParseEvtxTest.java [hunk 12]
```diff
@@ -0,0 +1,481 @@
...
+        verify(rootNodeHandler1).handle(rootNode1);
+        verify(rootNodeHandler1).close();
+        verify(rootNodeHandler2).handle(rootNode2);
+        verify(rootNodeHandler2).close();
+        verify(rootNodeHandler3).handle(rootNode3);
+        verify(rootNodeHandler3).close();
+    }
+
+    @Test
+    public void fileGranularityLifecycleTest() throws IOException, ParserConfigurationException, SAXException {
```
Comment (brosander) [inline nifi-nar-bundles/nifi-evtx-bundle/nifi-evtx-processors/src/test/java/org/apache/nifi/processors/evtx/ParseEvtxTest.java]: @mattyb149 lifecycle tests of the 3 different granularities using a doctored evtx file that has good chunks, bad chunk headers, and a malformed record

**Comment by mattyb149** [comment]
+1 LGTM, built and ran tests and contrib-check. Ran a NiFi flow with multiple EVTX files exercising all relationships and granularities.

Great contribution, thanks much!  Merging to master

**Comment by brosander** [comment]
changes merged upstream
